homeassistant/components/mqtt/client.py

```diff
@@ -644,10 +672,12 @@ def _mqtt_on_message(
         """Message received callback."""
         self.hass.add_job(self._mqtt_handle_message, msg)
 
-    @lru_cache(2048)
+    @lru_cache(None)  # pylint: disable=method-cache-max-size-none
     def _matching_subscriptions(self, topic: str) -> list[Subscription]:
         subscriptions: list[Subscription] = []
-        for subscription in self.subscriptions:
+        if topic in self._simple_subscriptions:
+            subscriptions.extend(self._simple_subscriptions[topic])
```

Comment:
Should there not be one match? after that we should not need to check the complex subscriptions?

Reply:
They can have multiple subscriptions for the same topic

Reply:
that is true indeed.